examples/warp_perspective.livemd

```diff
@@ -1,259 +1,170 @@
+    {"EVISION_ENABLE_CUDA", false},
+
+    # required when
+    # - `EVISION_ENABLE_CUDA` is `true`
+    # - and `EVISION_PREFER_PRECOMPILED` is `true`
+    #
+    # set `EVISION_CUDA_VERSION` to the version that matches
+    # your local CUDA runtime version
+    #
+    # current available versions are
+    # - 111
+    # - 114
+    # - 118
+    {"EVISION_CUDA_VERSION", "118"},
+
+    # require for Windows users when
+    # - `EVISION_ENABLE_CUDA` is `true`
+    # set `EVISION_CUDA_RUNTIME_DIR` to the directory that contains
+    # CUDA runtime libraries
+    {"EVISION_CUDA_RUNTIME_DIR", "C:/PATH/TO/CUDA/RUNTIME"}
+  ]
+)
+```
 
-  # required when 
-  # - `EVISION_ENABLE_CUDA` is `true`
-  # - and `EVISION_PREFER_PRECOMPILED` is `true`
-  #
-  # set `EVISION_CUDA_VERSION` to the version that matches 
-  # your local CUDA runtime version
-  #
-  # current available versions are
-  # - 111
-  # - 114
-  # - 118
-  {"EVISION_CUDA_VERSION", "118"},
+## Introduction
 
-  # require for Windows users when 
-  # - `EVISION_ENABLE_CUDA` is `true`
-  # set `EVISION_CUDA_RUNTIME_DIR` to the directory that contains
-  # CUDA runtime libraries
-  {"EVISION_CUDA_RUNTIME_DIR", "C:/PATH/TO/CUDA/RUNTIME"}
-])
-```
+This notebook will demonstrate how to perform [perspective transformation](https://en.wikipedia.org/wiki/3D_projection#Perspective_projection).
 
-<!-- livebook:{"output":true} -->
+It's useful to alias the module as something shorter when we make extensive use of the functions from certain modules.
 
-```
-:ok
+```elixir
+alias Evision, as: Cv
 ```
 
 ## Define Some Helper Functions
 
 ```elixir
 defmodule Helper do
-  def download!(url, save_as, overwrite? \\ false) do
-    unless File.exists?(save_as) do
-      Req.get!(url, http_errors: :raise, output: save_as, cache: not overwrite?)
-    end
-
-    :ok
+  def download!(url, overwrite \\ false) do
+    save_as = Path.join(System.tmp_dir!(), URI.encode_www_form(url))
+    unless File.exists?(save_as) || overwrite, do: Req.get!(url, output: save_as)
+    save_as
   end
-end
-```
 
-<!-- livebook:{"output":true} -->
+  @doc """
+  Returns the Euclidean norm.
 
-```
-{:module, Helper, <<70, 79, 82, 49, 0, 0, 10, ...>>, {:download!, 3}}
-```
-
-## Read the Test Image From File
-
-```elixir
-# Download the test image
-test_image_path = Path.join(__DIR__, "warp_perspective.png")
+  ## Examples
 
-Helper.download!(
-  "https://raw.githubusercontent.com/cocoa-xu/evision/main/test/testdata/warp_perspective.png",
-  test_image_path
-)
-
-# Read the test image
-%Evision.Mat{shape: {h, w, _}} = img = Evision.imread(test_image_path)
-```
-
-<!-- livebook:{"output":true} -->
-
-```
-%Evision.Mat{
-  channels: 3,
-  dims: 2,
-  type: {:u, 8},
-  raw_type: 16,
-  shape: {288, 277, 3},
-  ref: #Reference<0.391993608.626655254.103441>
-}
+    iex> Helper.hypot([3, 4])
+    5.0
+  """
+  def hypot([n | _] = numbers) when is_number(n) do
+    :math.sqrt(Enum.reduce(numbers, 0, &(&2 + &1 ** 2)))
+  end
```

Comment:
Just a really minor thing here. Should we perhaps still keep it as a lambda function? 

```elixir
# the hypot.(list(number())) function returns the Euclidean norm
hypot = fn n ->
  :math.sqrt(Enum.reduce(numbers, 0, &(&2 + &1 ** 2)))
end
```

I'm thinking that it feels more natural to do it this way (writing a lambda function where one needs it)? Also, I'm wondering if it would be better to keep functions in `Helper` for preparing the resources for the livebook?

What do you think?

Reply:
That makes sense!